### DISCUSSION TITLE
docker-slim build failure

### DISCUSSION THREAD

**Discussion by gurolakman** [slimtoolkit/slim#283] [Q&A]
Hello,

docker-slim build command fails consistently on my setup as follows. Any thoughts as to what may be wrong? Thanks!

docker-slim xray seems to work just fine as can be seen from below.

**% docker pull hello-world**
Using default tag: latest
latest: Pulling from library/hello-world
93288797bd35: Pull complete 
Digest: sha256:507ecde44b8eb741278274653120c2bf793b174c06ff4eaa672b713b3263477b
Status: Downloaded newer image for hello-world:latest
docker.io/library/hello-world:latest

**% docker-slim build hello-world**
docker-slim: message='join the Gitter channel to ask questions or to share your feedback' info='https://gitter.im/docker-slim/community'
docker-slim: message='join the Discord server to ask questions or to share your feedback' info='https://discord.gg/9tDyxYS'
docker-slim: message='Github discussions' info='https://github.com/docker-slim/docker-slim/discussions'
cmd=build info=param.http.probe message='using default probe' 
cmd=build state=started
cmd=build info=params target.type='image' target='hello-world' continue.mode='probe' rt.as.user='true' keep.perms='true' tags='' 
cmd=build state=image.inspection.start
cmd=build info=image size.human='9.1 kB' id='sha256:18e5af7904737ba5ef7fbbd7d59de5ebe6c4437907bd7fc436bf9b3ef3149ea9' size.bytes='9136' 
cmd=build info=image.stack index='0' name='hello-world:latest' id='sha256:18e5af7904737ba5ef7fbbd7d59de5ebe6c4437907bd7fc436bf9b3ef3149ea9' 
cmd=build state=image.inspection.done
cmd=build state=container.inspection.start
cmd=build info=container status='created' name='dockerslimk_21904_20220201011754' id='f8c0a24f8027fa7d850d506c95aa35b086822a8e02dd808ed041f66356ba2233' 
cmd=build info=cmd.startmonitor status='sent' 
cmd=build info=event.error status='received' data='SensorError{Op:sensor.fanotify.Run/fanapi.Initialize,Kind:call.error,Wrapped:{Type=syscall.Errno,Info=function not implemented,Line:43,File:github.com/docker-slim/docker-slim/pkg/app/sensor/monitors/fanotify/monitor.go}}' 
**cmd=build state=exited code=-124 component=container.inspector version=darwin|Transformer|1.37.3|latest|latest** 

**% docker-slim xray hello-world** 
docker-slim: message='join the Gitter channel to ask questions or to share your feedback' info='https://gitter.im/docker-slim/community'
docker-slim: message='join the Discord server to ask questions or to share your feedback' info='https://discord.gg/9tDyxYS'
docker-slim: message='Github discussions' info='https://github.com/docker-slim/docker-slim/discussions'
cmd=xray state=started
cmd=xray info=params target='hello-world' add-image-manifest='false' add-image-config='false' rm-file-artifacts='false' 
cmd=xray state=image.api.inspection.start
cmd=xray info=image id='sha256:18e5af7904737ba5ef7fbbd7d59de5ebe6c4437907bd7fc436bf9b3ef3149ea9' size.bytes='9136' size.human='9.1 kB' 
cmd=xray info=image.stack name='hello-world:latest' id='sha256:18e5af7904737ba5ef7fbbd7d59de5ebe6c4437907bd7fc436bf9b3ef3149ea9' instructions='2' message='see report file for details' index='0' 
cmd=xray state=image.api.inspection.done
cmd=xray state=image.data.inspection.start
cmd=xray info=image.data.inspection.save.image.start
cmd=xray info=image.data.inspection.save.image.end
cmd=xray info=image.data.inspection.process.image.start
cmd=xray info=image.data.inspection.process.image.end
cmd=xray state=image.data.inspection.done
cmd=xray info=image.package.details
cmd=xray info=layers.count value='1' 
cmd=xray info=layer.start
cmd=xray info=layer index='0' id='905047298562685366847c1265acd80e0df26f9d31df40b3bf6ae7f87dcdf0ca' path='905047298562685366847c1265acd80e0df26f9d31df40b3bf6ae7f87dcdf0ca/layer.tar' 
cmd=xray info=change.instruction snippet='COPY file:a79dd5bda1e77203401956a93401d3aef4...' all='COPY file:a79dd5bda1e77203401956a93401d3aef45221fc750295a4291896f3386f4f54 /' index='0:0' type='COPY' 
cmd=xray info=other.instructions count='1' 
cmd=xray info=other.instruction type='CMD' snippet='CMD ["/hello"]
' all='CMD ["/hello"]
' pos='0' index='0:1' 
cmd=xray info=layer.stats all_size.human='9.1 kB' all_size.bytes='9136' 
cmd=xray info=layer.stats object_count='1' 
cmd=xray info=layer.stats file_count='1' 
cmd=xray info=layer.stats max_file_size.human='9.1 kB' max_file_size.bytes='9136' 
cmd=xray info=layer.stats added_size.human='9.1 kB' added_size.bytes='9136' 
cmd=xray info=layer.change.summary deleted='0' modified='0' added='1' all='1' 
cmd=xray info=layer.objects.count value='1' 
cmd=xray info=layer.objects.top.start
A: mode=-rwxrwxr-x size.human='9.1 kB' size.bytes=9136 uid=0 gid=0 mtime='2021-09-24T01:20:44Z' H=[A:0] hash=84a28b6db3032962cb8d07012b015e40bcb520be '/hello'
cmd=xray info=layer.objects.top.end
cmd=xray info=layer.end
cmd=xray info=image.entry exe_path='/hello' exe_args='' 
cmd=xray info=image.entry.full_exe_path name='/hello' layer='0' 
cmd=xray state=completed
cmd=xray state=done
cmd=xray info=results artifacts.location='/tmp/docker-slim-state/.docker-slim-state/images/18e5af7904737ba5ef7fbbd7d59de5ebe6c4437907bd7fc436bf9b3ef3149ea9/artifacts' 
cmd=xray info=results artifacts.dockerfile.original='Dockerfile.fat' 
cmd=xray info=report file='slim.report.json' 
docker-slim: message='join the Gitter channel to ask questions or to share your feedback' info='https://gitter.im/docker-slim/community'
docker-slim: message='join the Discord server to ask questions or to share your feedback' info='https://discord.gg/9tDyxYS'
docker-slim: message='Github discussions' info='https://github.com/docker-slim/docker-slim/discussions'

**Comment by kcq** [comment]
@gurolakman what's the OS architecture and the environment where you are running docker-slim?

The `function not implemented` error for fanotify usually means it's a restricted environment...

What's strange is that you didn't get the `no exposed ports` error first because, by default, docker-slim expects to exposed ports in the target container and hello-world doesn't expose any

**Reply by gurolakman** [reply]
i was able to get docker-slim to work properly on an ubuntu server. so, it must be a thing w/ the m1 macbook. thanks for your help!

**Reply by kcq** [reply]
a similar problem with inotify on M1: https://github.com/docker/for-mac/issues/5321

there'll be an option to disable different engines in docker-slim soon, so it might be possible to get around it, but not sure if the other engine will not be restricted by the environment because it relies on another low level interface.

**Reply by gurolakman** [reply]
thanks a lot kyle! if & when this becomes available -- i'll be happy to test/validate it.

**Reply by gurolakman** [reply]
btw, the amount of reduction appears too good to be true! 

$ docker-slim build a3gw --http-probe=false
$ docker images | grep a3gw
a3gw.slim                                       latest        6d1f6b357a17   3 minutes ago        88.2MB
a3gw                                            latest        f5c00a133ccd   2 weeks ago          343MB

$ docker-slim build dmid --http-probe=false
$ docker images | grep dmid
dmid.slim                                       latest        02f28d82cb84   2 minutes ago        5.18MB
dmid                                            latest        1217c170b404   7 days ago           207MB

$ docker-slim build dmitestserver --http-probe=false
$ docker images | grep dmitestserver
dmitestserver.slim                              latest        ea90faafd7ad   7 minutes ago   3.56MB
dmitestserver                                   latest        b4ff504526af   7 months ago    632MB

$ docker-slim build dmitestclient --http-probe=false
$ docker images | grep dmitestclient
dmitestclient.slim                              latest        bd485e05f0cb   8 minutes ago    3.56MB
dmitestclient                                   latest        991d8d092702   7 months ago     632MB

disclaimer: we haven't tested/validated the slim images yet. does this look normal to you? thanks again.

**Comment by gurolakman** [comment]
<img width="210" alt="Screen Shot 2022-01-31 at 9 40 55 PM" src="https://user-images.githubusercontent.com/12719105/151905581-df257be5-e7a3-4689-90e5-705c3f90c599.png">

**Comment by randersson1000** [comment]
Docker slim doesn't seem to work on M1 processes, there is a thread about
it going on, seems strange to me but another interesting datapoint if we
can confirm it.

****
